Aliasing

bryteise opened the pull request.

matthewrsj commented on src/bundle.c:
```diff
@@ -1019,13 +1021,222 @@ static int install_bundles(struct list *bundles, struct list **subs, struct mani
+		ON_NULL_ABORT(l);
+
+		/* get the bundles associated with the alias
+                 * note a file with multiple copies of the same
```
You are mixing tabs and spaces in this patch.

bryteise replied:
Erf yes horribly.

bryteise replied:
Should be fixed.

matthewrsj commented on src/bundle.c:
```diff
@@ -1019,13 +1021,222 @@ static int install_bundles(struct list *bundles, struct list **subs, struct mani
+}
+
+/* Parse an alias definition file for a list of bundles
+ * it specifies
```
Can you add an example of what this file looks like in the comment?

bryteise replied:
Pretty trivial example added, let me know if you want more there.

matthewrsj replied:
nope that's perfect. Thank you.

matthewrsj commented on test/functional/bundleadd/add-alias-basic.bats:
```diff
@@ -0,0 +1,293 @@
+	remove_bundle -L "$TEST_NAME"/web-dir/10/Manifest.alias-bundle1
+	remove_bundle -L "$TEST_NAME"/web-dir/10/Manifest.alias-bundle2
+	remove_bundle -L "$TEST_NAME"/web-dir/10/Manifest.alias-bundle3
+        run sudo sh -c "rm -rf "$TARGETDIR"/usr/share/defaults/swupd/alias.d/"
```
tab and space mixing happening in test files as well.

bryteise replied:
Yicky and should be fixed.

otaviobp commented on src/bundle.c:
```diff
@@ -1095,22 +1272,28 @@ int install_bundles_frontend(char **bundles)
 	for (; *bundles; ++bundles) {
-		bundles_list = list_prepend_data(bundles_list, *bundles);
-		if (*bundles) {
+		struct list *alias_bundles = get_alias_bundles(aliases, *bundles);
```
I'm a bit concerned about if we should let system aliases override bundles that already exists.
A bundle emacs, for example could install a file creating an alias "vim emacs" and then vim bundle will never be installed. And there's nothing vim maintainers can do about that.

If a user override that, it's ok, but a package maintainer shouldn't

bryteise replied:
I'm inclined to say that is an content OS policy decision rather than update tool decision and that it can be tested against at that level if it is desired.

That said I may also be not excited to change this because doing a lookup to see if an alias defined is a bundle when creating definitions didn't seem particularly appealing to me when I was writing the code X(.

otaviobp replied:
> I'm inclined to say that is an content OS policy decision rather than update tool decision and that it can be tested against at that level if it is desired.

I took some time to think about that yesterday and I concluded you are right about that. There are so many other ways a package maintainer could create problems to other packages (create invalid configuration or library files, for example. The OS maintainers should be in charge to make sure this won't happen.

So, ignore my comment and leave this the way it is.

castulo commented on Makefile.am:
```diff
@@ -110,8 +110,11 @@ BATS_LOG_DRIVER = $(tap_driver)
-BATS = \
+BATS = test/functional/bundleadd/add-alias-basic.bats
+
+BBATS = \
```
I don't follow why you are doing this. Why are you adding this test twice? and why are you renaming the BATS tests as BBATS? is this something temporary so only your tests run while you figure out why they are failing?

bryteise replied:
Temporary for the test failures yes. Commit has been removed.

otaviobp commented on src/bundle.c:
```diff
@@ -1058,13 +1058,190 @@ static struct list *get_dir_files_sorted(char *path)
+	return bundles;
+}
+
+/* Parse an alias definition file for a list of bundles
```
Please, document too the order of override. And we need to document that in the manpages too in doc/ folder

bryteise replied:
Will do.

bryteise replied:
Do you want me to add an rst.in file to get the paths for system and user alias directories into the man pages?

bryteise replied:
I attempted to do this, hopefully everything looks reasonable now.

castulo commented on src/bundle.c:
```diff
@@ -1019,13 +1021,227 @@ static int install_bundles(struct list *bundles, struct list **subs, struct mani
+ * with their fullpath, returns NULL on error (errno set by
+ * opendir).
+ */
+static struct list *get_dir_files_sorted(char *path)
```
I feel like these function belongs to some other file, like helpers.c, or since there are many functions related to this maybe its own alias.c file

bryteise replied:
For the get_dir_files_sorted, putting it and lex_sort in helpers.c makes sense to me but because there isn't any other users outsdie of bundle.c I left it as is rather than polute the global namespace.

As for creating an alias.c, I think that would be fine but again as this only applies to bundle-add at the moment keeping them static seemed reasonable to me too.

It is not my project so just tell me whatever you all prefer and I'll do that =).

castulo replied:
ok, I'm ok leaving it as it is

castulo commented on test/functional/bundleadd/add-alias-basic.bats:
```diff
@@ -0,0 +1,310 @@
+#!/usr/bin/env bats
+
+# Author: <author name>
```
it would be nice if you add your contact information here in case someone has a question regarding these tests in the future

bryteise replied:
Ah thanks, I missed that.

otaviobp commented on src/bundle.c:
```diff
@@ -1058,13 +1058,190 @@ static struct list *get_dir_files_sorted(char *path)
+			}
+			iters = list_free_item(iters, free);
+
+			combined = list_prepend_data(combined, iteru->data);
```
this changes the order of user config files. If you have:

    /etc/swupd/alias.d/a1
    /etc/swupd/alias.d/b1
    /usr/share/defaults/swupd/alias.d/b1

We will look for aliases in the order:

    /etc/swupd/alias.d/b1
    /etc/swupd/alias.d/a1
    /usr/share/defaults/swupd/alias.d/b1

You can just remove the item from iters and concat system_alias_files with user_alias_files later.

bryteise replied:
Ah great catch, added test for this and will fix.

otaviobp commented on src/list.c:
```diff
@@ -280,7 +281,7 @@ struct list *list_deep_clone_strs(struct list *list)
 	item = list_tail(list);
 	while (item) {
-		clone = list_prepend_data(clone, strdup(item->data));
+		clone = list_prepend_data(clone, strdup_or_die(item->data));
```
+1

otaviobp commented on src/bundle.c:
```diff
@@ -1109,35 +1141,23 @@ static int lex_sort(const void *a, const void *b)
-	dir = opendir(path);
-	if (!dir) {
+	errno = 0;
+	bundles = get_dir_files_sorted(path);
```
I like the idea to always list bundles sorted.
+1 for that

castulo commented on test/functional/bundleadd/add-alias-basic.bats:
```diff
@@ -0,0 +1,308 @@
+
+	sudo sh -c "mkdir -p $TARGETDIR/etc/swupd/alias.d/"
+        run ls "$TARGETDIR/etc/swupd/"
+        assert_is_output --identical "alias.d"
```
it would be better if you use _assert_in_output_ instead of _assert_is_output_ since you are only interested into making sure that dir is there, because if we add some other content to that /etc/swupd directory your test is going to fail

bryteise replied:
That was only to help me debug the test failing later on.

bryteise commented on src/bundle.c:
```diff
@@ -1109,35 +1141,23 @@ static int lex_sort(const void *a, const void *b)
-	if (!dir) {
+	errno = 0;
+	bundles = get_dir_files_sorted(path);
+	if (errno) {
```
Well in theory somebody could just remove the files in /usr/share/clear/bundles and this would raise a false positive error. The later code doesn't have problems if bundles is null so I figured under reporting a potential issue (which would be obvious as it wouldn't show anything) isn't a big deal. I'm happy to change this though.

matthewrsj replied:
I agree with @bryteise here

bryteise replied:
So I'm not familiar with the case where opendir would succeed and readdir or closedir would fail given the same stream pointer and not having closed it already. I am happy to improve this though. Resetting errno before calling readdir and verifying errno is 0 after (otherwise free'ing the list) and then checking the return of closedir and freeing the list in case it has a non-zero return should resolve and unfree'd memory on program exit.

otaviobp replied:
sounds good

bryteise replied:
Fair enough, no reason to depend on the list functions not doing something weird in the future.

bryteise commented on configure.ac:
```diff
@@ -190,6 +190,22 @@ AS_IF(
+  [system_alias_path=${withval}],
+  [system_alias_path="/usr/share/defaults/swupd/alias.d/"]
+)
+AC_DEFINE_UNQUOTED([SYSTEM_ALIAS_PATH], ["${system_alias_path}"], [system-alias-path])
```
Why is a list of paths preferred for this? Could you explain how that design would work for the stateless behavior of the alias code too?

bryteise replied:
Hrm so I like the idea of having a config for users home but I'm not sure that would make too much sense given how swupd works today. But your point is totally reasonable I just think that it should have a use case associated with it first. I'll agreee with keeping this as is for now, thanks for the discussion!

bryteise commented on src/bundle.c:
```diff
@@ -1058,13 +1058,190 @@ static struct list *get_dir_files_sorted(char *path)
+ * alias1	b1	b2
+ * alias2	b3	b4
+ */
+static struct list *parse_alias_file(char *fullpath)
```
Okay. I'll move that code on, what did you think about the helper.c move that @castulo suggested?

otaviobp replied:
The perfect place for this file is sys.c, where we are grouping all system/filesystem related functions. But this didn't get merged yet, so leave it here and we can move to sys.c later.

bryteise replied:
So I moved the get_dir_files_sorted to helpers.c for now because I felt it was easier to work with since the function is used in the alias.c file now. I'll leave it for folks to move to sys.c when things are in better shape if that is okay?

otaviobp replied:
yes. it's totally fine.

bryteise commented on src/bundle.c:
```diff
@@ -1058,13 +1058,190 @@ static struct list *get_dir_files_sorted(char *path)
+		alias_content = list_prepend_data(alias_content, l);
+	}
+
+	fclose(afile);
```
Will do.

bryteise commented on src/bundle.c:
```diff
@@ -1058,13 +1058,190 @@ static struct list *get_dir_files_sorted(char *path)
+		struct alias_lookup *l = NULL;
+		char *c = NULL;
+
+		/* get the alias but ignore lines without bundles */
```
Okay.

bryteise commented on src/bundle.c:
```diff
@@ -1058,13 +1058,190 @@ static struct list *get_dir_files_sorted(char *path)
+		char *c = NULL;
+
+		/* get the alias but ignore lines without bundles */
+		c = strchr(line, '\t');
```
Size considerations of the file was pretty much the only reason for selecting this. I don't particularly see a reason to avoid more delimiters other than fewer options is safer than more for parsing.

otaviobp commented:
Another thing that came to my mind now about aliases. I don't know the future plans for alias, but we are not tracking the aliases that were installed, just the bundles. So, it would be nice to warn the user the list of bundles that he is installing from each alias. Something like:

    $ swupd bundle-add alias1
      Installing "vim emacs ..." bundles as 'required' by alias alias1

bryteise commented:
Hrm so I'm having second thoughts on printing out an installing specific for aliases. The issue I have is mostly increasing how ugly the code building the bundles_list_str is but I'm also a little unsure if I should be adding printing at that location given the output about installing bundles later on in the bundle_install code.

bryteise commented:
@otaviobp Is there anything more you'd like for me to do on the PR? I think I've addressed everything outside of the potential open on printing out what bundles are being installed via an alias.

bryteise commented on src/alias.c:
```diff
@@ -0,0 +1,220 @@
+		alias_content = list_prepend_data(alias_content, l);
+	}
+
+	if (errno) {
```
Okay, switching to the feof approach, forgot about that function thanks =).

bryteise commented on src/alias.c:
```diff
@@ -0,0 +1,220 @@
+		list_free_list_and_data(alias_content, free_alias_lookup);
+		alias_content = NULL;
+		(void)fclose(afile);
+	} else if (fclose(afile)) {
```
Okay, I'll closing the stream without checking errors.

bryteise commented on src/helpers.c:
```diff
@@ -1174,3 +1174,48 @@ int untar_full_download(void *data)
+		files = list_prepend_data(files, name);
+	}
+
+	if (errno) {
```
Okay.

bryteise commented on src/bundle.c:
```diff
@@ -22,6 +22,8 @@
 */
 
 #define _GNU_SOURCE
+#include <dirent.h>
```
Indeed, fixing.

bryteise commented on docs/swupd-alias.7.rst.in:
```diff
@@ -0,0 +1,57 @@
+=====
+
+Alias definitions can be defined multiple times in a file, if this happens the
+definition will be used.
```
Ugh that's a mess up X(.

bryteise commented:
@otaviobp I don't see any way to improve creating bundle_list_str aside from abstracting it behind a list function like reduce unfortunately. I added printing out the bundles added through alias and addressed the rest of the issues so the PR hopefully should be about ready now for merging.

castulo commented on test/functional/bundleadd/add-alias-basic.bats:
```diff
@@ -0,0 +1,333 @@
+
+}
+
+@test "ADD032: Add bundle without alias" {
```
I noticed that while this PR was being reviewed another PR that added a bundle-add test was merged, which was using ID ADD032.... so all these IDs need to be increased by 1 😓
